File: src/main/java/universalelectricity/compatibility/ModuleThermalExpansion.java
package universalelectricity.compatibility;

import cofh.api.energy.IEnergyContainerItem;
import cofh.api.energy.IEnergyHandler;
import net.minecraft.item.ItemStack;
import net.minecraftforge.common.ForgeDirection;
import universalelectricity.api.CompatibilityModule;
import universalelectricity.api.CompatibilityType;

/** @author Calclavia */
public class ModuleThermalExpansion extends CompatibilityModule
{
    @Override
    public long doReceiveEnergy(Object handler, ForgeDirection direction, long energy, boolean doReceive)
    {
        if (energy > 0)
        {
            double rf = energy * CompatibilityType.THERMAL_EXPANSION.ratio;
            if (rf > 0 && rf <= 1)
            {
                rf = 1;
            }
            return (long) (((IEnergyHandler) handler).receiveEnergy(direction, (int) rf, !doReceive) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
        }
        return 0;
    }

    @Override
    public long doExtractEnergy(Object handler, ForgeDirection direction, long energy, boolean doExtract)
    {
        if (energy > 0)
        {
            double rf = energy * CompatibilityType.THERMAL_EXPANSION.ratio;
            if (rf > 0 && rf <= 1)
            {
                rf = 1;
            }
            return (long) (((IEnergyHandler) handler).extractEnergy(direction, (int) rf, !doExtract) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
        }
        return 0;
    }

    @Override
    public boolean doIsHandler(Object obj)
    {
        return obj instanceof IEnergyHandler || obj instanceof IEnergyContainerItem;
    }

    @Override
    public boolean doCanConnect(Object obj, ForgeDirection direction, Object source)
    {
        return ((IEnergyHandler) obj).canInterface(direction);
    }

    @Override
    public long doChargeItem(ItemStack itemStack, long joules, boolean docharge)
    {
        if (itemStack.getItem() instanceof IEnergyContainerItem)
        {
            return (long) Math.floor(((IEnergyContainerItem) itemStack.getItem()).receiveEnergy(itemStack, (int) Math.ceil(joules * CompatibilityType.THERMAL_EXPANSION.ratio), !docharge) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
        }
        return 0;
    }

    @Override
    public long doDischargeItem(ItemStack itemStack, long joules, boolean doDischarge)
    {
        if (itemStack.getItem() instanceof IEnergyContainerItem)
        {
            return (long) Math.floor(((IEnergyContainerItem) itemStack.getItem()).extractEnergy(itemStack, (int) Math.ceil(joules * CompatibilityType.THERMAL_EXPANSION.ratio), !doDischarge) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
        }
        return 0;
    }

    @Override
    public ItemStack doGetItemWithCharge(ItemStack is, long energy)
    {
        return null;
    }

    @Override
    public boolean doIsEnergyContainer(Object obj)
    {
        return obj instanceof IEnergyHandler;
    }

    @Override
    public long doGetEnergy(Object obj, ForgeDirection from)
    {
        return (long) (((IEnergyHandler) obj).getEnergyStored(from) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
    }

    @Override
    public long doGetMaxEnergy(Object obj, ForgeDirection from)
    {
        return (long) (((IEnergyHandler) obj).getMaxEnergyStored(from) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
    }

    @Override
    public long doGetEnergyItem(ItemStack is)
    {
        if (is.getItem() instanceof IEnergyContainerItem)
        {
            return (long) (((IEnergyContainerItem) is.getItem()).getEnergyStored(is) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
        }
        return 0;
    }

    @Override
    public long doGetMaxEnergyItem(ItemStack is)
    {
        if (is.getItem() instanceof IEnergyContainerItem)
        {
            return (long) (((IEnergyContainerItem) is.getItem()).getMaxEnergyStored(is) * CompatibilityType.THERMAL_EXPANSION.reciprocal_ratio);
        }
        return 0;
    }
}

File: src/main/scala/universalelectricity/compatibility/module/ModuleThermalExpansion.scala
package universalelectricity.compatibility.module

import cofh.api.energy.{IEnergyContainerItem, IEnergyHandler}
import net.minecraft.item.ItemStack
import net.minecraftforge.common.util.ForgeDirection
import universalelectricity.compatibility.Compatibility

/**
 * Module for Thermal Expansion support
 * @author Calclavia
 */
object ModuleThermalExpansion extends Compatibility.CompatibilityModule("ThermalExpansion", "ThermalExpansion", "Redstone Flux", "RF", 50)
{
  def receiveEnergy(handler: AnyRef, direction: ForgeDirection, energy: Double, doReceive: Boolean): Double =
  {
    return (handler.asInstanceOf[IEnergyHandler]).receiveEnergy(direction, (energy * ratio).toInt, !doReceive) * reciprocal_ratio
  }

  def extractEnergy(handler: AnyRef, direction: ForgeDirection, energy: Double, doExtract: Boolean): Double =
  {
    return ((handler.asInstanceOf[IEnergyHandler]).extractEnergy(direction, (energy * ratio).toInt, !doExtract) * reciprocal_ratio)
  }

  def doIsHandler(obj: AnyRef): Boolean =
  {
    return obj.isInstanceOf[IEnergyHandler] || obj.isInstanceOf[IEnergyContainerItem]
  }

  def canConnect(obj: AnyRef, direction: ForgeDirection, source: AnyRef): Boolean =
  {
    return (obj.asInstanceOf[IEnergyHandler]).canConnectEnergy(direction)
  }

  def chargeItem(itemStack: ItemStack, joules: Double, docharge: Boolean): Double =
  {
    if (itemStack.getItem.isInstanceOf[IEnergyContainerItem])
    {
      return Math.floor((itemStack.getItem.asInstanceOf[IEnergyContainerItem]).receiveEnergy(itemStack, Math.ceil(joules * ratio).toInt, !docharge) * reciprocal_ratio)
    }
    return 0
  }

  def dischargeItem(itemStack: ItemStack, joules: Double, doDischarge: Boolean): Double =
  {
    if (itemStack.getItem.isInstanceOf[IEnergyContainerItem])
    {
      return Math.floor((itemStack.getItem.asInstanceOf[IEnergyContainerItem]).extractEnergy(itemStack, Math.ceil(joules * ratio).toInt, !doDischarge) * reciprocal_ratio)
    }
    return 0
  }

  def getItemWithCharge(is: ItemStack, energy: Double): ItemStack =
  {
    return null
  }

  def doIsEnergyContainer(obj: AnyRef): Boolean =
  {
    return obj.isInstanceOf[IEnergyHandler]
  }

  def getEnergy(obj: AnyRef, from: ForgeDirection): Double =
  {
    return ((obj.asInstanceOf[IEnergyHandler]).getEnergyStored(from) * reciprocal_ratio)
  }

  def getMaxEnergy(obj: AnyRef, from: ForgeDirection): Double =
  {
    return ((obj.asInstanceOf[IEnergyHandler]).getMaxEnergyStored(from) * reciprocal_ratio)
  }

  def getEnergyItem(is: ItemStack): Double =
  {
    if (is.getItem.isInstanceOf[IEnergyContainerItem])
    {
      return ((is.getItem.asInstanceOf[IEnergyContainerItem]).getEnergyStored(is) * reciprocal_ratio)
    }
    return 0
  }

  def getMaxEnergyItem(is: ItemStack): Double =
  {
    if (is.getItem.isInstanceOf[IEnergyContainerItem])
    {
      return ((is.getItem.asInstanceOf[IEnergyContainerItem]).getMaxEnergyStored(is) * reciprocal_ratio)
    }
    return 0
  }
}
